Spider/#7154 crawl maze

Leyart opened the pull request.

Leyart commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java:
```diff
@@ -133,71 +135,134 @@ public boolean parseResource(HttpMessage message, Source source, int depth) {
+        // formaction attributes override the action of the parent form, thus if no button with
+        // default action is provided, the default action should be removed
+
+        List<Element> formButtonElements =
```
I am sure I can avoid this double loop somehow, still looking for a solution

psiinon commented:
@Leyart you've jumped the gun on this one :P
We're only assigning one stackhawk-bounty issue to one person at a time, so you'll need to get #7159 merged first :)
I'll detail this in a blog post but I've not been able to find time to work on that yet...

Leyart commented:
> @Leyart you've jumped the gun on this one :P We're only assigning one stackhawk-bounty issue to one person at a time, so you'll need to get #7159 merged first :) I'll detail this in a blog post but I've not been able to find time to work on that yet...

I hope that the other one will be merged eventually and I had time that I could only spend at home. This is at the moment just a draft :)

Leyart commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java:
```diff
@@ -133,71 +135,125 @@ public boolean parseResource(HttpMessage message, Source source, int depth) {
-            getLogger().debug("Found new form with method: '" + method + "' and action: " + action);
 
-            // If no action, skip the form
-            if (action == null) {
```
I removed this check and wrapped the rest into a for loop over the form actions retrieved, the formatting make it scarier than it actually is

Leyart commented:
Should be in better shape than the previous one :) 
I will squash all commits once the review is done, but it is more practical to keep them separated per topic

psiinon reviewed (approved):
Those 5 tests pass locally with this change :)

thc202 commented on zap/src/test/java/org/zaproxy/zap/spider/parser/SpiderHtmlFormParserUnitTest.java:
```diff
@@ -777,6 +777,29 @@ void shouldUseRelativePathBaseHtmlUrlWhenParsingGetFormWithRelativeAction() {
     }
 
+    @Test
+    public void shouldUseButtonFormActionIfPresent() {
```
Should have package accessibility, same for the other test methods being added.

thc202 commented on zap/src/test/resources/org/zaproxy/zap/spider/parser/htmlform/FormWithFormactionButtons.html:
```diff
@@ -0,0 +1,32 @@
+<html>
+<head>
+    <meta charset="UTF-8">
+    <title>Form With Button Using Formaction - Spider HTML Form Parser</title>
```
Should have examples with:
 - Buttons with `form` attributes (in and outside forms);
 - Buttons with `formaction` but of `type` `reset` and `button`. Also, good to have some with explicit `submit` type and with invalid values.

thc202 replied:
Should have example containing invalid value and at least one button inside a form that has different `form` attribute.

Leyart commented on zap/src/test/resources/org/zaproxy/zap/spider/parser/htmlform/FormWithFormactionButtons.html:
```diff
@@ -0,0 +1,32 @@
+    <button formaction="http://example.org/formaction2">FormAction2</button>
+</form>
+
+<form action="http://example.org/form3" method="%%METHOD%%">
```
It was a test to verify that the correct formaction was picked up in case of no buttons, but since this test case is already covered by resources/org/zaproxy/zap/spider/parser/htmlform/FormNoSubmitField.html, there is no need for a duplicate indeed. I removed it

kingthorin commented:
Now conflicting.

kingthorin commented:
Nice! I meant to ask earlier if you needed help sorting it out but it looks like you've managed :grinning:

Leyart commented:
> Nice! I meant to ask earlier if you needed help sorting it out but it looks like you've managed 😀

I unfortunately pushed just before dropping off the train and didn't had time to fix it up to now, sorry about that :)

Leyart commented:
Help updated in https://github.com/zaproxy/zap-core-help/pull/435

thc202 commented:
The `formmethod` attribute should be handled as well (`formenctype` ideally, but since we don't for the form itself we don't for the button).

Leyart commented:
@thc202 Addressed the remaining concerns: improved the tests for the form buttons, including different form references and added support for formmethod with related tests. 
Not sure if I should add something additional to the formmethod ones.
I also updated the help in https://github.com/zaproxy/zap-core-help/pull/435 to address formmethod

Leyart commented:
> #7154

I have created https://github.com/zaproxy/zaproxy/issues/7197 to enhance form handling with the enctype attribute. 
I will take care of it but I would prefer to merge this PR first, since it is getting big enough already

Leyart commented:
> The `formmethod` attribute should be handled as well (`formenctype` ideally, but since we don't for the form itself we don't for the button).

I have created https://github.com/zaproxy/zaproxy/issues/7197 to enhance form handling with the enctype attribute. 
I will take care of it but I would prefer to merge this PR first, since it is getting big enough already

kingthorin commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java:
```diff
@@ -132,72 +135,189 @@ public boolean parseResource(HttpMessage message, Source source, int depth) {
+                            .collect(Collectors.toList()));
+        }
+
+        if (formButtonElements.size() > 0) {
```
`!formButtonElements.isEmpty()`

kingthorin commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java:
```diff
@@ -132,72 +135,189 @@ public boolean parseResource(HttpMessage message, Source source, int depth) {
+    }
+
+    /**
+     * Function that defines the allowed button types that should be considered in a form reset and
```
```suggestion
     * Function that defines the allowed button types that should be considered in a form, reset and
```

kingthorin commented on zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java:
```diff
@@ -132,72 +135,189 @@ public boolean parseResource(HttpMessage message, Source source, int depth) {
+    }
+
+    /**
+     * Defines the correct form method to use in case a button belongs to a form Valid form methods
```
```suggestion
     * Defines the correct form method to use in case a button belongs to a form. Valid form methods
```

kingthorin commented:
Thanks, I'm sure the javadoc tweaks seemed kinda nitpicky but to be honest I really had to think about them and the punctuation makes it easier to interpret (for my 2 cents anyway).

Leyart commented:
> Thanks, I'm sure the javadoc tweaks seemed kinda nitpicky but to be honest I really had to think about them and the punctuation makes it easier to interpret (for my 2 cents anyway).

I am not a native English speaker, actually far from that, so any nitpicking is more than welcomed :)

Leyart commented:
@kingthorin @psiinon @thc202 anything missing here? I may have time in the following days and then I may go offline for a while :)

kingthorin commented:
Sorry, I got into a routine and did that without checking. @psiinon @thc202 speak up if there was something else and I’ll tackle it.

psiinon commented:
It was all good from my side :)
